### ISSUE TITLE
Record package creates invalid files that are too small/corrupted

### ISSUE THREAD

**Issue by bloemy7** [llfbandit/record#394]
**Package version**: 5.1.2

**Environment**
 - OS: iOS, Android and web
 - Browser: chrome

**Describe the bug**

We are using the Record package in a production app and seeing daily dozens of recordings that upload to a server as corrupted files. This happens both for opus as well as aacLc.
The files will be super small: 5210 bytes, 4525 bytes, etc. and are simply corrupted - as in, they are not valid.
So, to decide on the encoder, we do this:
```
    String fileExtension = ".m4a";
    if (_recorder != null &&
        await _recorder!.isEncoderSupported(AudioEncoder.aacLc)) {
      codec = AudioEncoder.aacLc;
      fileExtension = ".m4a";
    } else if (_recorder != null &&
        await _recorder!.isEncoderSupported(AudioEncoder.opus)) {
      codec = AudioEncoder.opus;
      fileExtension = ".opus";
    } else if (_recorder != null &&
        await _recorder!.isEncoderSupported(AudioEncoder.pcm16bits)) {
      codec = AudioEncoder.pcm16bits;
      fileExtension = ".pcm";
    }
```
to try and cover all platforms.

And here is the rest of the config:
```
int bitRate = 32000; // 32 kbps
    int sampleRate = 16000; // 16 kHz
    RecordConfig config = RecordConfig(
        encoder: codec,
        numChannels: 1,
        bitRate: bitRate,
        sampleRate: sampleRate);
    await _recorder?.start(config, path: _filePath!);
```
Are we doing something obviously wrong here?

**To Reproduce**

Steps to reproduce the behavior: I don't have clear steps to reproduce since it doesn't happen for me - it happens in roughly 1-2% of the cases, but our app heavily depends on recordings so this is a real issue.

**Expected behavior**

Always generate a valid audio file.

**Additional context**

This happens on all platforms. Would love some help here, thanks a lot!

**Comment by llfbandit** [comment]
Thanks for the report.
But there's too many variables in this issue to determine what could the culprit.
Also the files you are opening are uploaded so we can't be sure of their integrity.

For the config, I would suggest to let bit rate and sample rate to default values and see how it goes.
On some platforms, those values may be adjusted as a best effort but this is not guaranted.

Finally, since you're mixing multiple platforms, I would suggest to ensure you are using the latest transitive dependencies.

**Comment by bloemy7** [comment]
@llfbandit Thanks for your answer! Just to clarify about the uploading part - I actually log the size error before the upload happens, so this is definitely unrelated to the upload (i.e I check the file size locally on the device and then upload, but if the size is too small I short circuit with an error).

I'll try changing the bit rate and sample rate to default in case it helps.

Regarding the dependencies, I'll also check. In case you have any other idea I would highly appreciate it.

**Comment by llfbandit** [comment]
Well to find the real issue you must change only one thing at a time, otherwise it will just hide it at best.
I would say by priority:

1. Transitive dependencies update and log somehow which platform/encoder pair triggers your check. We must restrict the scope to enable investigation.
2. Configuration.

**Comment by bloemy7** [comment]
@llfbandit thanks a lot. I'm adding logging right now so I'll be able to log that. Regarding transitive dependencies, what is a good way for me to check to make sure I'm using the most up to date ones? I'm definitely using the latest `record` package.

Thanks again.

**Comment by bloemy7** [comment]
@llfbandit I'm really trying to get to the bottom here because this is causing me real issues with customers out there. The weird part is that I was using another flutter library previously for recording, and I was seeing similar issues. This makes me think it could be related to my implementation, yet the way I'm doing it feels pretty standard, so I'm not sure either how that would make sense. I should start having some more information with logs tomorrow about what encoder/platform is causing issues but the fact it happens both on Web and on mobile already tells me it's not one encoder only, since it'll use opus on web but not on mobile.

**Comment by llfbandit** [comment]
Just run `flutter pub upgrade` and check that the latest versions are used. 
Safari should use aac not opus.

**Comment by bloemy7** [comment]
Thanks, I did that.

So part of the errors I'm seeing are:
`File handling error: ClientException: XMLHttpRequest error., uri=..../ea2b8a8d-9a74-4a99-94c1-b2bf52163ae7`
Those are very infrequent, but it basically is not able to retrieve the blog, right after we finish recording. This is on web only.

For the other ones it's still actually the file size that is too small:
<img width="185" alt="Screenshot 2024-09-05 at 7 11 06 PM" src="https://github.com/user-attachments/assets/6d9ff1ca-2831-4387-a02e-f4006b5bf2af">
It happens on web with both `aac` and `opus` (on Safari it uses `aac` automatically), so this seems totally independent of the platform/encoder, as it also happens on mobile.

Still trying to figure out the root cause here.

**Comment by bloemy7** [comment]
Hi @llfbandit , some of the errors that are being logged are as follows:
`File handling error: PathNotFoundException: Cannot retrieve length of file, path = '/data/user/0/appname/app_flutter/da87ca44-c737-42d7-b19f-ee5980eadf73_1726858219001.m4a' (OS Error: No such file or directory, errno = 2)`

Do you have any idea what could be causing this? I'm using the exact file URL that is returned to me by stopping the recording (and the same one that I give to it of course).
This only seems to happen in some cases, and only on Android. It's definitely an edge case but it happens multiple times a day for users in production. Does this happen to be a known error? This is when I try to manipulate the file right after I end recording.

Thanks!

**Comment by bloemy7** [comment]
@llfbandit finally, the other set of errors are actual m4a files that are generated but are corrupted - they won't play at all. Some of these are 8mb, others smaller, others larger. This happens on iOS only.
Do you have an idea of what could be causing this? I know these are not the most helpful bug reports but it's so strange since in general the recording works well for most cases, but in about ~5% of the times there are issues around corruption like this, which is too high to be something we can just ignore. Would love to hear from your experience what you think.

**Comment by bloemy7** [comment]
Hi @llfbandit, I'm coming back to you because we are still seeing this issue and as our usebase grows, it's happening more often. It's mainly on web, and it ends with recordings being 14kB and basically corrupted. I truly have no idea why this is happening and I'm running out of ideas. Is there any chance we can try some extra troubleshooting steps? The main issue is this is not something I can reproduce, but it's happening too often.

Thanks a lot

**Comment by Arkillon** [comment]
I'm seeing the same issue with this library. Using very similar configs & seems to happens mostly on android/ios for me.

Since my users grew in prod and are generating about 5000 audio files / day, I'm seeing exactly this issue with similar stats 1-2%.

Seeing both XMLHttpRequest error & PathNotFoundException as well. 

Will be investigating this issue & keeping an eye on this thread.

**Comment by bloemy7** [comment]
Thanks @Arkillon , please let me know if you find anything. I’m seeing it across all 3 platforms. It’s really becoming a major issue but I’m unable to find any pattern or solution so far. Let me know if you find anything. 
cc @llfbandit if you have any ideas on how we can investigate this. I’m happy to add any type of logging that can help us here. Thanks!

**Comment by bloemy7** [comment]
Hi @Arkillon let me know if I can contact you some way so we can maybe try to brainstorm about this together as it's really becoming a massive issue for our company as well. Thanks a lot!

**Comment by Arkillon** [comment]
@bloemy7 I Would be happy to brainstorm about this, feel free to reach out to me via discord "marc_paquin" or let me know if any other way is best for you.

**Comment by llfbandit** [comment]
Any update on your side?
What is weird here is you have the issue on 3 platforms while all 3 have completly different implementations.

**Comment by bloemy7** [comment]
@llfbandit it seems like @Arkillon might have found the root cause here, related to the recorder being in a scroll view, and thus getting "recycled" and losing state when scrolled out of view or tabs being switched around on web. We're still verifying this theory though !

**Comment by NoahJ-SK** [comment]
> Any update on your side? What is weird here is you have the issue on 3 platforms while all 3 have completly different implementations.

I'm in the same boat here. 2 different implementation, unexplained small corrupted recordings in ~2% of uses.

**Comment by darajava** [comment]
No, it's tough because the files we have on prod are sensitive and we need permission. I don't get too many actual angry users reporting in so my best guess is just that it's a super short recording that this lib can't handle.

**Comment by bloemy7** [comment]
@darajava I'm having roughly the same issues. Ultimately it's a very low percentage of people, but those people are quite angry, so it's definitely something I'm trying to fix. It's very hard to figure out what might be causing it.

**Comment by ppamorim** [comment]
@ospaarmann iOS simulator is pretty much unreliable and broken. Use a real device.

We ran into a related issue in our production app (`record: ^6.2.0, AudioEncoder.pcm16bits, sampleRate: 16000, mono, startStream`).

What we found on iOS Simulator:

The native `RecorderStreamDelegate.swift` correctly creates an `AVAudioConverter` from `audioEngine.inputNode.inputFormat(forBus: 0)` (Mac hardware, typically 48 kHz) to the requested output format (16 kHz). The conversion code is correct. However, the simulator's audio subsystem delivers roughly one-third of the expected bytes per second ~10 660 B/s instead of 32 000 B/s for 16 kHz mono 16-bit PCM. This means frames are being dropped somewhere between the `installTap` callback and the Dart stream listener. The result is files that are valid structurally but contain far less audio content than the recording duration implies.

Real devices are not affected. We confirmed this on a physical iPhone where byte throughput matches the requested sample rate exactly.

**Comment by llfbandit** [comment]
Closing this.
This is old and fuzzy now. Feel free to re-open an issue if you confirm that still happen with latest version 7.1+.
